llvm/lib/Transforms/Vectorize/VPlanPatternMatch.h

```diff
@@ -66,6 +66,27 @@ struct specificval_ty {
 
 inline specificval_ty m_Specific(const VPValue *VPV) { return VPV; }
 
+/// Stores a reference to the VPValue *, not the VPValue * itself,
+/// thus can be used in commutative matchers.
+template <typename Class> struct deferredval_ty {
```

Comment:
Is it necessary to match over anything else other than VPValue? Doesn't the name deferredval_ty imply that we're matching a value?

Reply:
The template is to allow matching both VPValue and const VPValue, which ties into your review comment below. If we don't think we'll ever need to match a const VPValue I can remove that overload and untemplate this function.

Reply:
IIRC, we don't have a const matcher for m_VPValue(), and considering that m_Deferred() is supposed to be used along with m_VPValue(), I think it's safe to remove the const variant, and un-templatize this.

Reply:
Good point about there not being a const m_VPValue(), I've untemplated it now